src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs

```diff
@@ -0,0 +1,319 @@
+            AlertThreshold = (double)temp;
+            Host.CheckDecode(AlertThreshold >= 0 && AlertThreshold <= 1);
+        }
+
+        private protected SrCnnTransformBase(SrCnnArgumentBase args, string name, IHostEnvironment env)
+            : this(args.WindowSize, args.InitialWindowSize, args.Source, args.Name,
+                  name, env, args.BackAddWindowSize, args.LookaheadWindowSize, args.AvergingWindowSize, args.JudgementWindowSize, args.Threshold)
+        {
+        }
+
+        private protected override void SaveModel(ModelSaveContext ctx)
+        {
+            Host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel();
+
+            Host.Assert(WindowSize > 0);
+            Host.Assert(InitialWindowSize == WindowSize);
+            Host.Assert(BackAddWindowSize > 0);
+            Host.Assert(LookaheadWindowSize > 0);
+            Host.Assert(AvergingWindowSize > 0);
+            Host.Assert(JudgementWindowSize > 0);
+            Host.Assert(AlertThreshold >= 0 && AlertThreshold <= 1);
+
+            base.SaveModel(ctx);
+            ctx.Writer.Write((byte)BackAddWindowSize);
+            ctx.Writer.Write((byte)LookaheadWindowSize);
+            ctx.Writer.Write((byte)AvergingWindowSize);
+            ctx.Writer.Write((byte)JudgementWindowSize);
+            ctx.Writer.Write((byte)AlertThreshold);
+        }
+
+        internal override IStatefulRowMapper MakeRowMapper(DataViewSchema schema) => new Mapper(Host, this, schema);
+
+        internal sealed class Mapper : IStatefulRowMapper
+        {
+            private readonly IHost _host;
+            private readonly SrCnnTransformBase<TInput, TState> _parent;
+            private readonly DataViewSchema _parentSchema;
+            private readonly int _inputColumnIndex;
+            private readonly VBuffer<ReadOnlyMemory<Char>> _slotNames;
+            private SrCnnStateBase State { get; set; }
+
+            public Mapper(IHostEnvironment env, SrCnnTransformBase<TInput, TState> parent, DataViewSchema inputSchema)
+            {
+                Contracts.CheckValue(env, nameof(env));
+                _host = env.Register(nameof(Mapper));
+                _host.CheckValue(inputSchema, nameof(inputSchema));
+                _host.CheckValue(parent, nameof(parent));
+
+                if (!inputSchema.TryGetColumnIndex(parent.InputColumnName, out _inputColumnIndex))
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", parent.InputColumnName);
+
+                var colType = inputSchema[_inputColumnIndex].Type;
+                if (colType != NumberDataViewType.Single)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", parent.InputColumnName, "Single", colType.ToString());
+
+                _parent = parent;
+                _parentSchema = inputSchema;
+                _slotNames = new VBuffer<ReadOnlyMemory<char>>(2, new[] { "Alert".AsMemory(), "Raw Score".AsMemory(),
```

Comment:
This is passing `2` in the for length, but the array has `3` elements. Is this intended?

Reply:
Fixed. Thanks!